src/services/TestResultsService.ts

```diff
@@ -128,38 +139,83 @@ export class TestResultsService {
-          this.setAuditDetails(newTestResult, oldTestResult, msUserDetails);
-          newTestResult.testResultId = uuidv4();
-          return this.testResultsDAO.updateTestResult(newTestResult, oldTestResult)
-              .then((data) => {
-                return newTestResult;
-              }).catch((error) => {
-                throw new HTTPError(500, error.message);
-              });
-        }).catch((error) => {
-          throw new HTTPError(error.statusCode, error.body);
-        });
+      .then((result) => {
+        const response: ITestResultData = {Count: result.Count, Items: result.Items};
+        const testResults = this.checkTestResults(response);
+        const oldTestResult = this.getTestResultToArchive(testResults, payload.testResultId);
+        const newTestResult: ITestResult = cloneDeep(oldTestResult);
+        newTestResult.testVersion = TEST_VERSION.CURRENT;
+        oldTestResult.testVersion = TEST_VERSION.ARCHIVED;
+        mergeWith(newTestResult, payload);
+        this.setAuditDetails(newTestResult, oldTestResult, msUserDetails);
+        newTestResult.testResultId = uuidv4();
+        return this.testResultsDAO.updateTestResult(newTestResult, oldTestResult)
+            .then((data) => {
+              return newTestResult;
+            }).catch((error) => {
+              throw new HTTPError(500, error.message);
+            });
+      }).catch((error) => {
+        throw new HTTPError(error.statusCode, error.body);
+      });
+  }
+
+  public validateTestTypes(testResult: ITestResult) {
+    const validationErrors = [];
+    let validation: ValidationResult<any> | any;
+    for (const testType of testResult.testTypes) {
+      const context = {isPassed: testType.testResult, isSubmitted: testResult.testStatus};
+      if (TEST_TYPES_GROUP1.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup1.validate(testType, {context});
+      } else if (TEST_TYPES_GROUP2.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup2.validate(testType, {context});
+      } else if (TEST_TYPES_GROUP3_4_5_10.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup3And4And5And10.validate(testType, {context});
+      } else if (TEST_TYPES_GROUP6_7_8.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup6And7And8.validate(testType, {context});
+      } else if (TEST_TYPES_GROUP9_11.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup9And11.validate(testType, {context});
+      } else if (TEST_TYPES_GROUP12_13.includes(testType.testTypeId)) {
+        validation = testTypesSchemaGroup12And13.validate(testType, {context});
+      } else {
+        validation = {
+          error: {
+            details: [{message: "Unknown testTypeId"}]
+          }
+        };
+      }
+      if (validation.error) {
+        validationErrors.push(validation.error.details.map((detail: { message: string; }) => {
```

Comment:
We seemed reusing this pattern across 2 other different places, maybe this can be extracted into a separate function? Possibly a separate util file if it is also used in other services?

Reply:
will do, it's not used elsewhere, will extract to a function